src/lseq/ident.rs

```diff
@@ -255,39 +262,55 @@ impl<A: Actor> IdentGen<A> {
     }
 }
 
-#[cfg(test)]
 impl<A: quickcheck::Arbitrary + Actor> quickcheck::Arbitrary for Identifier<A> {
     fn arbitrary<G: quickcheck::Gen>(g: &mut G) -> Identifier<A> {
-        Identifier {
-            path: Vec::<(u64, Option<A>)>::arbitrary(g),
+        let max_depth = 27; // TODO: where does this come from?
+        let min_depth = 1;
+        let chosen_depth = u32::arbitrary(g) % (max_depth - min_depth) + min_depth;
+        let mut path = Vec::new();
+        for depth in 0..chosen_depth {
+            let i = u64::arbitrary(g) % (2u64.pow(INITIAL_BASE + depth) + 1);
+            path.push((i, Option::arbitrary(g)));
+        }
+
+        // our last entry can not use 0 as an index since this will not allow us to insert
+        // anything before this identifier
+        if path.last().unwrap().0 == 0 {
+            path.pop();
+            let i = 1 + u64::arbitrary(g) % 2u64.pow(INITIAL_BASE + chosen_depth);
+            assert_ne!(i, 0);
+            path.push((i, Option::arbitrary(g)));
+        };
+
+        Self { path }
+    }
+
+    fn shrink(&self) -> Box<dyn Iterator<Item = Self>> {
+        if self.path.len() == 1 {
+            Box::new(std::iter::empty())
+        } else {
+            let mut path = self.path.clone();
+            path.pop();
+            Box::new(std::iter::once(Self { path }))
         }
     }
 }
 
 #[cfg(test)]
 mod test {
     use super::*;
-    use quickcheck::{quickcheck, Arbitrary, Gen, TestResult};
-
-    fn valid_identifier(id: &Identifier<u32>) -> bool {
```

Comment:
I've moved this logic into the Arbitrary::arbitrary implementation. We don't generate invalid identifiers anymore so we don't need this validation.